plugins/parade-lspcon/fu-parade-lspcon-device.c

```diff
@@ -0,0 +1,832 @@
+	if (!fu_parade_lspcon_poll_register (self, REG_ADDR_SPI_CTL, SPI_CTL_TRIGGER, 0, error))
+		return FALSE;
+
+	/* read SR value */
+	return fu_parade_lspcon_read_register (self, REG_ADDR_RD_FIFO, value, error);
+}
+
+/* poll the flash status register for operation completion */
+static gboolean
+fu_parade_lspcon_flash_wait_ready (FuParadeLspconDevice *self, GError **error)
+{
+	g_autoptr(GTimer) timer = g_timer_new ();
+
+	do {
+		guint8 status_register;
+		if (!fu_parade_lspcon_flash_read_status (self, &status_register, error))
+			return FALSE;
+
+		/* BUSY bit clears on completion */
+		if ((status_register & 1) == 0)
+			return TRUE;
+
+		/* flash operations generally take between 1ms and 4s; polling
+		 * at 1000 Hz is still quite responsive and not overly slow */
+		g_usleep (G_TIME_SPAN_MILLISECOND);
+	} while (g_timer_elapsed (timer, NULL) <= 10.0);
+
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_TIMED_OUT,
+			     "flash did not become ready within 10 seconds");
+	return FALSE;
+}
+
+static gboolean
+fu_parade_lspcon_flash_write (FuParadeLspconDevice *self,
+	     guint32 base_address,
+	     GBytes *data,
+	     GError **error)
+{
+	FuDevice *device = FU_DEVICE (self);
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	const guint8 unlock_writes[] = { 0xaa, 0x55, 0x50, 0x41, 0x52, 0x44 };
+	gsize data_len = g_bytes_get_size (data);
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* address must be 256-byte aligned */
+	g_return_val_if_fail ((base_address & 0xFF) == 0, FALSE);
+	g_debug ("flash write %" G_GSIZE_FORMAT " bytes at %#x",
+		 g_bytes_get_size (data), base_address);
+
+	/* unlock map writes by writing the magic sequence */
+	for (gsize i = 0; i < sizeof(unlock_writes); i++) {
+		if (!fu_parade_lspcon_write_register (self,
+						      REG_ADDR_MAP_WRITE,
+						      unlock_writes[i],
+						      error))
+			return FALSE;
+	}
+
+	/* reset clt2SPI, required before write */
+	if (!fu_parade_lspcon_write_register (self, REG_ADDR_CLT2SPI, 0x20, error))
+		return FALSE;
+	g_usleep (100 * G_TIME_SPAN_MILLISECOND);
+	if (!fu_parade_lspcon_write_register (self, REG_ADDR_CLT2SPI, 0, error))
+		return FALSE;
+
+	chunks = fu_chunk_array_new_from_bytes (data, base_address, 0, 256);
+	for (gsize i = 0; i < chunks->len; i++) {
+		FuChunk *chunk = g_ptr_array_index (chunks, i);
+		guint32 address = fu_chunk_get_address (chunk);
+		guint32 chunk_size = fu_chunk_get_data_sz (chunk);
+		guint8 write_data[257] = { 0x0 };
+		g_autoptr(FuParadeLspconI2cAddressGuard) guard = NULL;
+
+		/* map target address range in page 7 */
+		if (!fu_parade_lspcon_map_page (self, address, error))
+			return FALSE;
+
+		/* write data to page 7 memory window */
+		guard = fu_parade_lspcon_i2c_address_guard_new (self,
+								I2C_ADDR_PAGE7,
+								error);
+		/* page write is prefixed with an offset:
+		 * we always start from offset 0 */
+		write_data[0] = 0;
+		memcpy(write_data + 1, fu_chunk_get_data (chunk), chunk_size);
```

Comment:
Can you use `fu_memcpy_safe` here please.

Reply:
Done.